src/error-handling/exercise.rs

```diff
@@ -12,212 +12,101 @@
-        }
-    }
-}
-
-#[derive(Debug, Error)]
-enum ParserError {
-    #[error("Tokenizer error: {0}")]
-    TokenizerError(#[from] TokenizerError),
-    #[error("Unexpected end of input")]
-    UnexpectedEOF,
-    #[error("Unexpected token {0:?}")]
-    UnexpectedToken(Token),
-    #[error("Invalid number")]
-    InvalidNumber(#[from] std::num::ParseIntError),
-}
-
-fn parse(input: &str) -> Result<Expression, ParserError> {
-    let mut tokens = tokenize(input);
-
-    fn parse_expr<'a>(
-        tokens: &mut Tokenizer<'a>,
-    ) -> Result<Expression, ParserError> {
-        let tok = tokens.next().ok_or(ParserError::UnexpectedEOF)??;
-        let expr = match tok {
-            Token::Number(num) => {
-                let v = num.parse()?;
-                Expression::Number(v)
-            }
-            Token::Identifier(ident) => Expression::Var(ident),
-            Token::Operator(_) => return Err(ParserError::UnexpectedToken(tok)),
-        };
-        // Look ahead to parse a binary operation if present.
-        Ok(match tokens.next() {
-            None => expr,
-            Some(Ok(Token::Operator(op))) => Expression::Operation(
-                Box::new(expr),
-                op,
-                Box::new(parse_expr(tokens)?),
-            ),
-            Some(Err(e)) => return Err(e.into()),
-            Some(Ok(tok)) => return Err(ParserError::UnexpectedToken(tok)),
-        })
-    }
-
-    parse_expr(&mut tokens)
-}
-
-fn main() -> anyhow::Result<()> {
-    let expr = parse("10+foo+20-30")?;
-    println!("{expr:?}");
-    Ok(())
-}
-// ANCHOR_END: solution
+#[derive(PartialEq, Eq, Debug)]
+struct DivideByZeroError;
+// ANCHOR_END: types
 
 /*
-// ANCHOR: panics
-fn tokenize(input: &str) -> Tokenizer {
-    return Tokenizer(input.chars().peekable());
-}
-
-struct Tokenizer<'a>(Peekable<Chars<'a>>);
-
-impl<'a> Tokenizer<'a> {
-    fn collect_number(&mut self, first_char: char) -> Token {
-        let mut num = String::from(first_char);
-        while let Some(&c @ '0'..='9') = self.0.peek() {
-            num.push(c);
-            self.0.next();
-        }
-        Token::Number(num)
-    }
-
-    fn collect_identifier(&mut self, first_char: char) -> Token {
-        let mut ident = String::from(first_char);
-        while let Some(&c @ ('a'..='z' | '_' | '0'..='9')) = self.0.peek() {
-            ident.push(c);
-            self.0.next();
+// ANCHOR: eval
+// The original implementation of the expression evaluator. Update this to
+// return a `Result` and produce an error when dividing by 0.
+fn eval(e: Expression) -> i64 {
+    match e {
+        Expression::Op { op, left, right } => {
+            let left = eval(*left);
+            let right = eval(*right);
+            match op {
+                Operation::Add => left + right,
+                Operation::Sub => left - right,
+                Operation::Mul => left * right,
+                Operation::Div => if right != 0 {
```

Comment:
`src/pattern-matching/exercise.rs` does not contain this conditional (it just relies on `/` to panic).

I think that's OK, but maybe claiming it's "the original" in the comment above is inaccurate enough to bother folks (like me)

Reply:
Yeah true it's not exactly the same as the solution to the earlier exercise, I deliberately added in the panic to show students where exactly the error case is. Personally I don't think that'd be too confusing for students, but let me add a speaker note calling that out to teachers so they can mention it if students are confused.